Improve process output

julienfalque opened the pull request:
This PR changes the process output to look like PHPUnit's one:
* limits line length to 80 characters;
* displays the progression percentage before each newline so you know whether you have time to get a coffee 😉;

![Output](https://cloud.githubusercontent.com/assets/1736542/23835879/40c9cae4-076f-11e7-8d0f-3831d1222e54.png)

julienfalque commented on composer.json:
```diff
@@ -17,7 +17,7 @@
         "ext-tokenizer": "*",
         "sebastian/diff": "^1.1",
-        "symfony/console": "^2.3 || ^3.0",
+        "symfony/console": "^2.4 || ^3.0",
```
The `BufferedOutput` class used in test was introduced in 2.4.

SpacePossum replied:
2.3 is supported to May this year (http://symfony.com/roadmap?version=2.3#checker), would be nice if we could keep it (especially if the up is only for test). However if this is a lot of work, than up'ping is :+1:  with me

keradus replied:
bumping is fine

sstok replied:
Note that 2.4 is EOL so you'd properly want 2.7 (if it's still active) or 2.8 😃

keradus replied:
As I do always answer to you @sstok, if we could work properly on both 2.8 and 2.4 no need for dropping 2.4.
I saw a lot of legacy, monolithic projects that updates their codebase to have better CS and this kind of change would disallow them to install the tool as dev-dependency and require them to install it as phar file.

julienfalque replied:
As @SpacePossum said, the class is only used for tests and is not very complex, maybe we can just write our own?

keradus replied:
if we want to use feature of 2.4, let us bump to 2.4.

GrahamCampbell commented on src/Console/Command/FixCommand.php:
```diff
@@ -184,7 +184,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         $progressOutput = $showProgress && $stdErr
-            ? new ProcessOutput($stdErr, $this->eventDispatcher)
+            ? new ProcessOutput($stdErr, $this->eventDispatcher, count($resolver->getFinder()))
```
This `count` is very expensive. It forces the iterator to go all the way through.

julienfalque replied:
On the project (445 files), it takes ~0.05s (basic `microtime()` test). Is it really that expensive?

I tried to store the files in an array and use it for both counting and running the fixers. Don't know if it reduces execution time but it consumes 4M memory.

SpacePossum replied:
The cost depends on the Finder implementation used and on what set it is used, so the numbers provided give some indication but not much IMHO.
It is the same discussion as here https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1408
I'm not in favor (but not a strong opinion) to always count ahead, but as an opt-in I think this looks nice (or provide an opt-out back to the current way of dots).
Happy to hear more opinions though!

keradus replied:
@julienfalque , imagine finder that checks for content, like all files that do not contain `ignore CS`.
but I do agree that this is very edge case.

storing outcome of finder in memory may be difficult if the project is extra large. we could try approach like `if ($count < SOME_LIMIT) { $finder = ArrayIterator(iterator_to_array($finder)); }`
The, all the rest work out of the box, count is cheap and we don't iterate over disc for searching for the files twice if there are not that much files.

SpacePossum replied:
To would enforce that the finders used must be rewindable for when after first iteration the limit is reached.

keradus replied:
if it's not rewindable we could convert it to array anyway (even if it would be big memory-consuming array)
next argument to have 2 `ProcessOutput` implementations

GrahamCampbell commented on src/Console/Output/ProcessOutput.php:
```diff
@@ -68,6 +85,20 @@ public function onFixerFileProcessed(FixerFileProcessedEvent $event)
+        $end = $this->processedFiles === $this->files;
+        if (0 === $symbolsOnCurrentLine || $end) {
+            $this->output->writeln(sprintf(
+                '%s %'.strlen((string) $this->files).'d / %d (%3d%%)',
```
Bit of a hack to cast this to a string, in order to then use it later in sprintf, uncast. Abstractly, we can't be sure the `%d` gives something of the same length as your string cast did.

julienfalque replied:
> Abstractly, we can't be sure the %d gives something of the same length as your string cast did.

What values would give such difference?

SpacePossum replied:
only downside I see is that is done in the constructor (76) already. I would remove it from the constructor and reuse the `strlen((string) $this->files)` in a var.

julienfalque replied:
Why in the destructor?

SpacePossum replied:
Because it is already there?

julienfalque replied:
There is nothing in the destructor except the removal of the event listener. Did you mean `onFixerFileProcessed`?

SpacePossum replied:
oh sorry! I misread the GH interface, on the callback sounds good to me :+1:

keradus commented:
what about 2 reporters - `txt` and `simple_txt` (current one) ?

GrahamCampbell commented:
> Do we want to have that kind of progress indicator

Maybe. Would be cool to have it as optional.

julienfalque commented:
What about adding an option like `--show-progress` that would enable showing the number of files and the progress percentage? By default lines length would still be limited to 80 characters, without further informations.

keradus commented:
IMO, we should:
1. add new option for command
- `new InputOption('show-progress', '', InputOption::VALUE_REQUIRED, 'Specify progress indicator.')` (no default value !)
- possible values are: `none, run-in, estimating` (means none, current way, new way); feel free to propose other naming; would be good to document the difference between them
- if option is not provided in CLI command run, it shall fallback to `ConfigInterface::getHideProgress : bool`, which means `run-in` or `none`
- resolving which `ProgressOutput` shall be use needs to be done via `ConfigurationResolver::getProgress`, which will return `ProcessOutputInterface` now instead of boolean flag

2. Update config
- `Config::getProgress() : ProcessOutputInterface` shall be added, but without touching interface
- in `ConfigurationResolver` we should detect if the method is there, if so we shall use it, if not we shall use `ConfigurationResolver::getProgress`

3. new `ProgressOutput` class
- as finder may be any `iterable`, it may not be rewindable, having that in mind, if we gonna use new progress class in given run I would conditionally transform finder to regular array, so it will become countable (not all `iterables` are `countable` !). I know it will increase memory usage for big projects, we should mention it in user documentation of progress types.

julienfalque commented:
Sounds good to me. Should `ConfigInterface::getHideProgress` be deprecated?

keradus commented:
if you want to deprecate sth you need to say that this is abandoned or shall be done other way. here, it's the second case. but to say "do it in another way", you need to provide that way. new interface perhaps

julienfalque commented:
`Config::getProgress()` could be part of a new interface that extends `ConfigInterface`.

julienfalque commented:
New option added, but I didn't implement it exactly like you suggested: the `ProcessOutputInterface` instance is still created in the `FixCommand`.

Also, with `run-in` option output is still limited to 80 characters per line but without the numbers so it has no impact on memory.

keradus commented:
`run-in` name had sense when it would represent current output.

but you did changed it (thus the name may not be the best as you suggested, as the format is different) - not sure is it good to change current output instead of providing different progress type.

name should follow the representation. I am very against `full`, as it's full in what way? use full screen, `run-in` use only 40% of full type, or what? and then, even if one would be able to provide reasoning why `full` may be an OK name for current types, then I will add one-liner progress `bar`, which works completly different than `full`/`estimating`

julienfalque commented:
Updated README file and reverted `run-in` to original single-line rendering.

> I am very against full, as it's full in what way?

Sure, doesn't really make sense, let's keep current options 👍

keradus commented on README.rst:
```diff
@@ -187,6 +187,18 @@ Risky rule is a rule, which could change code behaviour. By default no risky rul
+
+* ``none``: disables progress output;
+* ``run-in``: simple single-line progress output;
+* ``evaluating``: multiline progress output with number of files and percentage on each line.
```
please mention memory issue with `evaluating` for big projects

keradus commented on src/Console/Output/ProcessOutput.php:
```diff
@@ -52,11 +52,31 @@
+     */
+    private $symbolsPerLine;
+
+    public function __construct(OutputInterface $output, EventDispatcher $dispatcher, $nbFiles = null)
```
typehinting of last param is missing, you need to add a docblock

keradus replied:
also, please remove default value

keradus commented on src/Console/Output/ProcessOutput.php:
```diff
@@ -52,11 +52,31 @@
+
+        if (null !== $nbFiles) {
+            $this->files = $nbFiles;
+            $this->symbolsPerLine = 80 - strlen((string) $this->files) * 2 - 11;
```
for me those are magic values, 80, 11, 2 ;)

keradus commented on src/Console/Output/ProcessOutput.php:
```diff
@@ -68,6 +88,27 @@ public function onFixerFileProcessed(FixerFileProcessedEvent $event)
+
+        if (null !== $this->files) {
+            $symbolsOnCurrentLine = $this->processedFiles % $this->symbolsPerLine;
+            $end = $this->processedFiles === $this->files;
```
boolean vars should be named with prefix like `is` or `has`

keradus commented on src/Console/Output/ProcessOutput.php:
```diff
@@ -68,6 +88,27 @@ public function onFixerFileProcessed(FixerFileProcessedEvent $event)
         $status = self::$eventStatusMap[$event->getStatus()];
         $this->output->write($this->output->isDecorated() ? sprintf($status['format'], $status['symbol']) : $status['symbol']);
+
+        ++$this->processedFiles;
```
could be done only when `this->files` is not null

keradus commented on src/Console/Output/ProcessOutput.php:
```diff
@@ -52,11 +52,40 @@
 
-    public function __construct(OutputInterface $output, EventDispatcher $dispatcher)
+    /**
+     * @var int
```
`?int`

keradus commented:
Thank you @julienfalque.
